packages/thirdweb/src/react/web/ui/prebuilt/thirdweb/BuyDirectListingButton/index.tsx

```diff
@@ -0,0 +1,189 @@
+ * and then pass it to a <TransactionButton />
+ *
+ * Since it uses the TransactionButton, it can take in any props that can be passed
+ * to the [`TransactionButton`](https://portal.thirdweb.com/references/typescript/v5/TransactionButton)
+ *
+ *
+ * @param props
+ * @returns <TransactionButton />
+ *
+ * @example
+ * ```tsx
+ * import { BuyDirectListingButton } from "thirdweb/react";
+ *
+ * <BuyDirectListingButton
+ *   contractAddress="0x..." // contract address of the marketplace v3
+ *   chain={...} // the chain which the marketplace contract is deployed on
+ *   client={...} // thirdweb client
+ *   listingId={100n} // the listingId or the item you want to buy
+ *   quantity={1n} // optional - see the docs to learn more
+ * >
+ *   Buy NFT
+ * </BuyDirectListingButton>
+ * ```
+ */
+export function BuyDirectListingButton(props: BuyDirectListingButtonProps) {
+  const {
+    contractAddress,
+    listingId,
+    children,
+    chain,
+    client,
+    quantity,
+    payModal,
+  } = props;
+  const defaultPayModalMetadata = payModal ? payModal.metadata : undefined;
+  const account = useActiveAccount();
+  const contract = getContract({
+    address: contractAddress,
+    client,
+    chain,
+  });
+
+  const { data: payMetadata } = useReadContract(getPayMetadata, {
+    contract,
+    listingId,
+    queryOptions: {
+      enabled: !defaultPayModalMetadata,
+    },
+  });
+
+  const { mutateAsync } = useSendAndConfirmTransaction();
+
+  async function prepareBuyTransaction() {
+    if (!account) {
+      throw new Error("No account detected");
+    }
+    const [listing, { getApprovalForTransaction }, { buyFromListing }] =
+      await Promise.all([
+        getListing({
+          contract,
+          listingId,
+        }),
+        import(
+          "../../../../../../extensions/erc20/write/getApprovalForTransaction.js"
+        ),
+        import(
+          "../../../../../../extensions/marketplace/direct-listings/write/buyFromListing.js"
+        ),
+      ]);
+    if (!listing) {
+      throw new Error(`Could not retrieve listing with ID: ${listingId}`);
+    }
+
+    // For ERC721 listing, always hard-code to 1n
+    const _quantity =
+      listing.asset.type === "ERC721" ? 1n : quantity || listing.quantity;
+
+    if (listing.asset.type === "ERC1155") {
+      if (_quantity < 1n) {
+        throw new Error("Invalid `quantity` - expected at least 1n");
+      }
+      if (_quantity > listing.quantity) {
+        throw new Error(
+          `quantity exceeds available amount. Available: ${listing.quantity.toString()}`,
+        );
+      }
+    }
+
+    const buyTx = buyFromListing({
+      contract,
+      listingId,
+      quantity: _quantity,
+      recipient: account?.address || "",
+    });
+
+    const approveTx = await getApprovalForTransaction({
+      transaction: buyTx,
+      account,
+    });
+
+    if (approveTx) {
```

Comment:
Kind of weird we execute a transaction inside the prepare but I guess that's the limitation of TransactionButton

Reply:
Yeah.. I talked to Joaquim about making a change to the TB, so that we can "chain" multiple transactions.

But for now we think this should do